### PR TITLE
Allow dynamic addition of new BatchSystem

### PR DESCRIPTION
**Pull request by evan-wehi**
This changes allows the registration of new implementations of AbstractBatchSystem. There is a registry for factories to create new BatchSystems (toil.batchSystems.registry) and a seperate registry for adding command line options (toil.batchSystems.options). The registries are pre-populated with the exiting BatchSystem implementations. The reason for separating options and factories is that command line options can be present even if the dependencies for the batch system are not. (Arguably, you might not want command line options present if the can't be used). 

The motivating use case is for sites (or even individual users) to be able to add BatchSystems that can implement local policies or to implement BatchSystems that are not generally useful. It also decouples the concrete BatchSystem implementations from `common.py`.

New BatchSystem can be added programmatically as follows:

```
from toil.batchSystems.registry import addBatchSystemFactory
from toil.batchSystems.options import addOptionsDefinition
    
def drmaaBatchSystemFactory():
    from wehi_pipeline.batchSystems.drmaaBatchSystem import DrmaaBatchSystem
    return DrmaaBatchSystem
    
addBatchSystemFactory('drmaa', drmaaBatchSystemFactory)
       
def addOptions(addOptionFn):
    addOptionFn("--jobQueue", dest="jobQueue", default=None,
            help=("A job queue (used by the DRMAA batch system"))
        
addOptionsDefinition(addOptions)
```

### CODE REVIEW

**Comment by cket** [comment]
Jenkins, test this please

**Comment by cket** [comment]
Jenkins, ok to test

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
There we go, sorry about that. I just had to tweak some setting in our Jenkins config to authorize myself to test PRs from outside our organization

File: src/toil/batchSystems/abstractBatchSystem.py [hunk 1]
```diff
@@ -184,6 +184,16 @@ def getRescueBatchJobFrequency(cls):
         raise NotImplementedError()
 
 
+    @staticmethod
```
Comment (cket) [inline src/toil/batchSystems/abstractBatchSystem.py]: What about \@abstractmethod instead? Seems to fit that pattern better

File: src/toil/batchSystems/abstractBatchSystem.py [hunk 2]
```diff
@@ -184,6 +184,16 @@ def getRescueBatchJobFrequency(cls):
...
+    def setOptions(setOption):
+        """
+        Process command line or configuration options relevant to this batch system
+        
+        :param common.Config: the Config object
```
Comment (cket) [inline src/toil/batchSystems/abstractBatchSystem.py]: I think either this comment or the parameter name is misleading.

File: .gitignore [hunk 3]
```diff
@@ -1,16 +0,0 @@
-*.pyc
```
Comment (cket) [inline .gitignore]: I don't think we don't want to delete this file ;)

**Comment by evan-wehi** [comment]
Hi Christopher,

thanks for the feedback. The idea of setOptions is an overridable static, a concept that doesn’t seem to exist in python (or java, which I’m more familiar with). It’s not abstract since most classes with be happy with a base implementation A classmethod may be a better option. That said, when I went to add meaningful comments, that interface seems overly obscure so I’ll have another crack at it.

Evan.
--
Evan Thomas
Research Computing Scientist
Computational Biology
The Walter and Eliza Hall Institute of Medical Research
Internal: 2323 | Mobile: +61423000246 | Skype: evanathomas | Email: thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>




On 13 Dec. 2016, at 6:16 am, Christopher Ketchum <notifications@github.com<mailto:notifications@github.com>> wrote:


@cket requested changes on this pull request.

@evan-wehi<https://github.com/evan-wehi> thanks for the PR! The PR isn't building correctly on Jenkins because we renamed our 'test' target. If you rebase/merge from master that problem should fix itself.

I also left a couple comments on a couple small issues, I'll take a closer look once we get the tests running on this PR!

________________________________

In src/toil/batchSystems/abstractBatchSystem.py<https://github.com/BD2KGenomics/toil/pull/1374#pullrequestreview-12526663>:

> @@ -184,6 +184,16 @@ def getRescueBatchJobFrequency(cls):
         raise NotImplementedError()


+    @staticmethod


What about @abstractmethod instead? Seems to fit that pattern better

________________________________

In src/toil/batchSystems/abstractBatchSystem.py<https://github.com/BD2KGenomics/toil/pull/1374#pullrequestreview-12526663>:

> @@ -184,6 +184,16 @@ def getRescueBatchJobFrequency(cls):
         raise NotImplementedError()


+    @staticmethod
+    def setOptions(setOption):
+        """
+        Process command line or configuration options relevant to this batch system
+
+        :param common.Config: the Config object


I think either this comment or the parameter name is misleading.

________________________________

In .gitignore<https://github.com/BD2KGenomics/toil/pull/1374#pullrequestreview-12526663>:

> @@ -1,16 +0,0 @@
-*.pyc


I don't think we don't want to delete this file ;)

**Comment by evan-wehi** [comment]
Hi Christopher,

I’ve merged from your repo but there seems to file(s) missing:

Traceback (most recent call last):
  File "/Volumes/thomas.e/dev/micronuclei/src/toil_test.py", line 88, in <module>
    main()
  File "/Volumes/thomas.e/dev/micronuclei/src/toil_test.py", line 84, in main
    t.start(mj)
  File "/Volumes/thomas.e/dev/toil/src/toil/common.py", line 625, in start
    return self._runMainLoop(rootJobGraph)
  File "/Volumes/thomas.e/dev/toil/src/toil/common.py", line 901, in _runMainLoop
    logProcessContext(self.config, logger)
  File "/Volumes/thomas.e/dev/toil/src/toil/__init__.py", line 147, in logProcessContext
    from toil.version import version
ImportError: No module named version


Thanks,
Evan.
--
Evan Thomas
Research Computing Scientist
Computational Biology
The Walter and Eliza Hall Institute of Medical Research
Internal: 2323 | Mobile: +61423000246 | Skype: evanathomas | Email: thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>




On 13 Dec. 2016, at 10:29 am, Evan Thomas <thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>> wrote:

Hi Christopher,

thanks for the feedback. The idea of setOptions is an overridable static, a concept that doesn’t seem to exist in python (or java, which I’m more familiar with). It’s not abstract since most classes with be happy with a base implementation A classmethod may be a better option. That said, when I went to add meaningful comments, that interface seems overly obscure so I’ll have another crack at it.

Evan.
--
Evan Thomas
Research Computing Scientist
Computational Biology
The Walter and Eliza Hall Institute of Medical Research
Internal: 2323 | Mobile: +61423000246 | Skype: evanathomas | Email: thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>




On 13 Dec. 2016, at 6:16 am, Christopher Ketchum <notifications@github.com<mailto:notifications@github.com>> wrote:


@cket requested changes on this pull request.

@evan-wehi<https://github.com/evan-wehi> thanks for the PR! The PR isn't building correctly on Jenkins because we renamed our 'test' target. If you rebase/merge from master that problem should fix itself.

I also left a couple comments on a couple small issues, I'll take a closer look once we get the tests running on this PR!

________________________________

In src/toil/batchSystems/abstractBatchSystem.py<https://github.com/BD2KGenomics/toil/pull/1374#pullrequestreview-12526663>:

> @@ -184,6 +184,16 @@ def getRescueBatchJobFrequency(cls):
         raise NotImplementedError()


+    @staticmethod


What about @abstractmethod instead? Seems to fit that pattern better

________________________________

In src/toil/batchSystems/abstractBatchSystem.py<https://github.com/BD2KGenomics/toil/pull/1374#pullrequestreview-12526663>:

> @@ -184,6 +184,16 @@ def getRescueBatchJobFrequency(cls):
         raise NotImplementedError()


+    @staticmethod
+    def setOptions(setOption):
+        """
+        Process command line or configuration options relevant to this batch system
+
+        :param common.Config: the Config object


I think either this comment or the parameter name is misleading.

________________________________

In .gitignore<https://github.com/BD2KGenomics/toil/pull/1374#pullrequestreview-12526663>:

> @@ -1,16 +0,0 @@
-*.pyc


I don't think we don't want to delete this file ;)

**Comment by evan-wehi** [comment]
Ok - this seems to be due to the fact that I run Toil directly from the src directory. I take it version.py is built at some point but in previous testing it wasn’t called so I didn’t notice it wasn’t there. Sorry for the basic question, but how should I run toil when developing? (I’m not a python developer.)

Thanks,
Evan.
--
Evan Thomas
Research Computing Scientist
Computational Biology
The Walter and Eliza Hall Institute of Medical Research
Internal: 2323 | Mobile: +61423000246 | Skype: evanathomas | Email: thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>




On 14 Dec. 2016, at 1:46 pm, Evan Thomas <thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>> wrote:

Hi Christopher,

I’ve merged from your repo but there seems to file(s) missing:

Traceback (most recent call last):
  File "/Volumes/thomas.e/dev/micronuclei/src/toil_test.py", line 88, in <module>
    main()
  File "/Volumes/thomas.e/dev/micronuclei/src/toil_test.py", line 84, in main
    t.start(mj)
  File "/Volumes/thomas.e/dev/toil/src/toil/common.py", line 625, in start
    return self._runMainLoop(rootJobGraph)
  File "/Volumes/thomas.e/dev/toil/src/toil/common.py", line 901, in _runMainLoop
    logProcessContext(self.config, logger)
  File "/Volumes/thomas.e/dev/toil/src/toil/__init__.py", line 147, in logProcessContext
    from toil.version import version
ImportError: No module named version


Thanks,
Evan.
--
Evan Thomas
Research Computing Scientist
Computational Biology
The Walter and Eliza Hall Institute of Medical Research
Internal: 2323 | Mobile: +61423000246 | Skype: evanathomas | Email: thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>




On 13 Dec. 2016, at 10:29 am, Evan Thomas <thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>> wrote:

Hi Christopher,

thanks for the feedback. The idea of setOptions is an overridable static, a concept that doesn’t seem to exist in python (or java, which I’m more familiar with). It’s not abstract since most classes with be happy with a base implementation A classmethod may be a better option. That said, when I went to add meaningful comments, that interface seems overly obscure so I’ll have another crack at it.

Evan.
--
Evan Thomas
Research Computing Scientist
Computational Biology
The Walter and Eliza Hall Institute of Medical Research
Internal: 2323 | Mobile: +61423000246 | Skype: evanathomas | Email: thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>




On 13 Dec. 2016, at 6:16 am, Christopher Ketchum <notifications@github.com<mailto:notifications@github.com>> wrote:


@cket requested changes on this pull request.

@evan-wehi<https://github.com/evan-wehi> thanks for the PR! The PR isn't building correctly on Jenkins because we renamed our 'test' target. If you rebase/merge from master that problem should fix itself.

I also left a couple comments on a couple small issues, I'll take a closer look once we get the tests running on this PR!

________________________________

In src/toil/batchSystems/abstractBatchSystem.py<https://github.com/BD2KGenomics/toil/pull/1374#pullrequestreview-12526663>:

> @@ -184,6 +184,16 @@ def getRescueBatchJobFrequency(cls):
         raise NotImplementedError()


+    @staticmethod


What about @abstractmethod instead? Seems to fit that pattern better

________________________________

In src/toil/batchSystems/abstractBatchSystem.py<https://github.com/BD2KGenomics/toil/pull/1374#pullrequestreview-12526663>:

> @@ -184,6 +184,16 @@ def getRescueBatchJobFrequency(cls):
         raise NotImplementedError()


+    @staticmethod
+    def setOptions(setOption):
+        """
+        Process command line or configuration options relevant to this batch system
+
+        :param common.Config: the Config object


I think either this comment or the parameter name is misleading.

________________________________

In .gitignore<https://github.com/BD2KGenomics/toil/pull/1374#pullrequestreview-12526663>:

> @@ -1,16 +0,0 @@
-*.pyc


I don't think we don't want to delete this file ;)

**Comment by cket** [comment]
@evan-wehi you should first run 'make develop' from the project root and then you should be able to run Toil from anywhere. That should generate the missing version.py file

**Comment by evan-wehi** [comment]
Thank you - seems to be working now.


--
Evan Thomas
Research Computing Scientist
Computational Biology
The Walter and Eliza Hall Institute of Medical Research
Internal: 2323 | Mobile: +61423000246 | Skype: evanathomas | Email: thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>




On 15 Dec. 2016, at 4:54 am, Christopher Ketchum <notifications@github.com<mailto:notifications@github.com>> wrote:


@evan-wehi<https://github.com/evan-wehi> you should first run 'make develop' from the project root and then you should be able to run Toil from anywhere. That should generate the missing version.py file

**Comment by evan-wehi** [comment]
Because I have 2 PRs in play, I've moved this to another branch. I'm closing this PR and I'll submit another PR incorporating the requested changes.

**Pull request closed without merging**